babel/numbers.py

```diff
@@ -416,15 +429,54 @@ def format_decimal(
                                  the format pattern. Defaults to `True`.
     :param group_separator: Boolean to switch group separator on/off in a locale's
                             number format.
+    :param format_type: Format to use (`None`, "short" or "long"). The standard decimal
+                        format is `None`. Defaults to `None`.
+    :param compact_fraction_digits: Number of fraction digits to use in "short" or "long"
+                                    format. Defaults to `0`. If this is set to
+                                    a value greater than `0`, the `decimal_quantization`
+                                    will be treated as `False`.
     """
     locale = Locale.parse(locale)
-    if not format:
-        format = locale.decimal_formats.get(format)
+    if format_type in ("short", "long"):
+        number, format = _get_compact_format(number, format_type, locale, compact_fraction_digits)
+        # if compact_fraction_digits is set, we don't want to truncate the fraction digits
+        decimal_quantization = False if compact_fraction_digits > 0 else decimal_quantization
+    elif not format:
+        format = locale.decimal_formats.get(format_type)
     pattern = parse_pattern(format)
     return pattern.apply(
         number, locale, decimal_quantization=decimal_quantization, group_separator=group_separator)
 
 
+def _get_compact_format(number, format_type, locale, compact_fraction_digits=0):
+    """Returns the number after dividing by the unit and the format pattern to use.
+    The algorithm is described here:
+    https://www.unicode.org/reports/tr35/tr35-45/tr35-numbers.html#Compact_Number_Formats.
+    """
+    format = None
+    compact_format = locale.compact_decimal_formats[format_type]
+    for magnitude in sorted([int(m) for m in compact_format["other"]], reverse=True):
+        if abs(number) >= magnitude:
+            # check the pattern using "other" as the amount
+            format = compact_format["other"][str(magnitude)]
+            pattern = parse_pattern(format).pattern
+            # if the pattern is "0", we do not divide the number
+            if pattern == "0":
+                break
+            # otherwise, we need to divide the number by the magnitude but remove zeros
+            # equal to the number of 0's in the pattern minus 1
+            number = number / (magnitude / (10 ** (pattern.count("0") - 1)))
+            # round to the number of fraction digits requested
+            number = round(number, compact_fraction_digits)
+            # if the remaining number is singular, use the singular format
+            plural_form = locale.plural_form(abs(number))
+            plural_form = plural_form if plural_form in compact_format else "other"
+            format = compact_format[plural_form][str(magnitude)]
+            break
+    format = format if format is not None else locale.decimal_formats.get(None)
```

Comment:
I think this could be clearer as a regular `if`.
```suggestion
    if format is None:  # Did not find a format, fall back.
        format = locale.decimal_formats.get(None)
```
or similar..?